File: 1008-construct-binary-search-tree-from-preorder-traversal/1008-construct-binary-search-tree-from-preorder-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    // int preidx = -1;
    TreeNode* bstFromPreorder(vector<int>& preorder) {
        // 1. push root to stack
        // 2. if preord[idx] < top then it is the left node & poush 
        // 3. if preord[idx]>top keep popping the values till we    have the first nowhich is greater then for prev pop connect it as   right node & push it
        return generateTree(preorder);
    }

    TreeNode* generateTree(vector<int>& preorder){
        stack<TreeNode*> stack; 
        TreeNode* root=new TreeNode(preorder.front());
        stack.push(root);
        for(int i=1;i<preorder.size();i++){
            TreeNode* node = new TreeNode(preorder[i]);
            if(stack.size() && stack.top()->val>node->val){
                stack.top()->left = node; 
                stack.push(node);
            }
            else {
                TreeNode* prev=NULL;
                while(stack.size() && stack.top()->val<node->val){
                    prev=stack.top();
                    stack.pop();
                }
                if(prev) prev->right=node;
                stack.push(node);
            }
        }
        return root;
    }

};

File: 1008-construct-binary-search-tree-from-preorder-traversal/1008-construct-binary-search-tree-from-preorder-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def __init__(self): 
        self.preidx = 0 
    

    def bstFromPreorder(self, preorder: List[int]) -> Optional[TreeNode]:
        # each node gets pushed into stack if less than top 
        # if greater pop till you find the next greater
        # it would be right node of prev 
        # if st empty then the prev val only regardlles 
        # first val is root 
        st=[] 
        root = TreeNode(preorder[0])
        st.append(root) 
        for i in range(1,len(preorder)): 
            currnode = TreeNode(preorder[i])
            if st and st[-1].val>currnode.val: 
                st[-1].left = currnode
                st.append(currnode) 
            else: 
                prev =None 
                while st and st[-1].val<currnode.val: 
                    prev=st.pop()
                prev.right = currnode
                st.append(currnode) 

        return root 